internal/events/loop.go

```diff
@@ -5,158 +5,94 @@ import (
+// FIXME(pleshakov): better document the side effects and how to prevent and mitigate them.
+// So when the EventLoop have 100 saved events, it is better to process them at once rather than one by one.
 type EventLoop struct {
-	cfg EventLoopConfig
+	eventCh <-chan interface{}
+	logger  logr.Logger
+	handler EventHandler
 }
 
 // NewEventLoop creates a new EventLoop.
-func NewEventLoop(cfg EventLoopConfig) *EventLoop {
-	return &EventLoop{cfg: cfg}
+func NewEventLoop(eventCh <-chan interface{}, logger logr.Logger, handler EventHandler) *EventLoop {
+	return &EventLoop{
+		eventCh: eventCh,
+		logger:  logger,
+		handler: handler,
+	}
 }
 
 // Start starts the EventLoop.
-// The method will block until the EventLoop stops:
-// - if it stops because of an error, the Start will return the error.
-// - if it stops normally, the Start will return nil.
+// This method will block until the EventLoop stops, which will happen after the ctx is closed.
+//
+// FIXME(pleshakov). Ensure that when the Gateway starts, the first time it generates configuration for NGINX,
+// it has a complete view of the cluster resources. For example, when the Gateway processes a Gateway resource
+// with a listener with TLS termination enabled (the listener references a TLS Secret), the Gateway knows about the secret.
+// This way the Gateway will not produce any incomplete transient configuration at the start.
 func (el *EventLoop) Start(ctx context.Context) error {
+	// The current batch.
+	var batch EventBatch
+	// handling tells if any batch is currently being handled.
+	var handling bool
+	// handlingDone is used to signal the completion of handling a batch.
+	handlingDone := make(chan struct{})
+
+	handle := func(ctx context.Context, batch EventBatch) {
+		el.handler.HandleEventBatch(ctx, batch)
+		handlingDone <- struct{}{}
+	}
+
+	// Note: at any point of time, no more than one batch is currently being handled.
+
+	// The event loop
 	for {
 		select {
 		case <-ctx.Done():
-			// although we always return nil, Start must return it to satisfy
-			// "sigs.k8s.io/controller-runtime/pkg/manager".Runnable
+			// Wait for the completion if a batch is being handled.
+			if handling {
+				<-handlingDone
+			}
 			return nil
-		case e := <-el.cfg.EventCh:
-			el.handleEvent(ctx, e)
+		case e := <-el.eventCh:
+			// Add the event to the current batch.
+			batch = append(batch, e)
+
+			// FIXME(pleshakov): Log more details about the event like resource GVK and ns/name.
+			el.logger.Info(
+				"added an event to the current batch",
+				"type", fmt.Sprintf("%T", e),
+				"total", len(batch),
+			)
+
+			// Handle the current batch if no batch is being handled.
+			if !handling {
+				go handle(ctx, batch)
+				batch = make([]interface{}, 0)
+				handling = true
+			}
+		case <-handlingDone:
+			handling = false
+
+			// Handle the current batch if it has at least one event.
+			if len(batch) > 0 {
+				go handle(ctx, batch)
+				batch = make([]interface{}, 0)
```

Comment:
I'm trying to come up with a relevant enough benchmark test.

But making an entirely new, zero-length slice seems unnecessary and potentially inefficient.

We could trade runtime size for performance I think. If we double buffer we can always add to one slice while reading from the other. When it's time to process swap the buffers and reset the old one. They retain their max used size but reduce memory operations once "warmed up".

Reply:
I added a FIXME. Would that be suffice for now?

Reply:
The isolated (possibly contrived) tests I have now show double buffers to be a clear winner. Let's discuss offline.

`package slicereuse

const depth = 256

func populateCopyRecreate(s *[]int) {
  for j := 0; j < depth; j++ { 
    *s = append(*s, j)
  } 

  current := *s
  var _ = current
  *s = make([]int, 0)
}

func populateBuffer(s *[]int) {
  for j := 0; j < depth; j++ { 
    *s = append(*s, j)
  } 

  current := *s
  var _ = current
  *s = (*s)[:0]
} 

/*
func populateCopyRecreate(s []int) {
  for j := 0; j < depth; j++ {
    s = append(s, j)
  }

  current := s
  var _ = current
  s = make([]int, 0)
}

func populateBuffer(s []int) {
  for j := 0; j < depth; j++ {
    s = append(s, j)
  }

  current := s
  var _ = current
  s = s[:0]
} 
*/

...

package slicereuse

import (
  "testing"
)

func BenchmarkPopulateCopyRecreate(b *testing.B) {
  s := make([]int, 0)
  for n := 0; n < b.N; n++ {
    // fast
    populateCopyRecreate(&s)
    // slow
    //populateCopyRecreate(s)
  }

  //fmt.Printf("len s: %v - cap s: %v\n", len(s), cap(s))
}

func BenchmarkPopulateDoubleBuffer(b *testing.B) {
  s1 := make([]int, 0)
  s2 := make([]int, 0)
  for n := 0; n < b.N; n++ {
    if n%2 == 0 {
      // fast
      populateBuffer(&s1)
      // slow
      //populateBuffer(s1)
    } else {
      // fast
      populateBuffer(&s2)
      // slow
      //populateBuffer(s2)
    }
  }

  //fmt.Printf("len s1: %v - cap s1: %v\n", len(s1), cap(s1))
  //fmt.Printf("len s2: %v - cap s2: %v\n", len(s2), cap(s2))
}
`

Results:
`goos: linux
goarch: amd64
pkg: slicereuse
cpu: Intel(R) Xeon(R) CPU E3-1505M v5 @ 2.80GHz
BenchmarkPopulateCopyRecreate-8   	  956547	      1254 ns/op
BenchmarkPopulateDoubleBuffer-8   	 2987036	       404.2 ns/op
PASS
ok  	slicereuse	3.759s
`

Reply:
Go with a FIXME and technical debt issue because of time constraints.

Please update the comment to:

> FIXME(pleshakov): Making an entirely new buffer is inefficient and multiplies memory operations. Use a double-buffer approach - create two buffers and exchange them between the producer and consumer routines. NOTE: pass-by-reference, and reset buffer to length 0, but retain capacity.